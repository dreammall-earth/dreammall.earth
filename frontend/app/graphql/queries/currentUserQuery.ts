import { gql } from 'graphql-tag'

export const currentUserQuery = gql`
  query {
    currentUser {
      id
      referenceId
      name
      username
      introduction
      availability
      details {
        id
        category
        text
      }
      social {
        id
        type
        link
      }
      table {
        id
        name
        public
        users {
          id
          role
          name
          username
        }
      }
    }
  }
`
