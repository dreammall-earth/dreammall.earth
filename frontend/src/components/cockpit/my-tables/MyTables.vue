<template>
  <CockpitCard>
    <template #header>
      <div class="header">
        <h2>{{ $t('cockpit.myTables.header') }}</h2>
        <button
          class="add-table bg-primary px-3 py-1 d-flex align-center justify-center"
          @click="addTable"
        >
          <v-icon icon="mdi mdi-plus" />
          {{ $t('cockpit.myTables.addTable') }}
        </button>
      </div>
    </template>
    <template #default>
      <ul class="list">
        <TableItem v-for="table in tables" :key="table.id" v-bind="table" />
      </ul>
      <div v-if="tables.length === 0">
        {{ $t('cockpit.myTables.noTables') }}
      </div>
    </template>
  </CockpitCard>
</template>

<script lang="ts" setup>
import { storeToRefs } from 'pinia'
import { computed } from 'vue'

import CockpitCard from '#components/cockpit/cockpit-card/CockpitCard.vue'
import useModal from '#components/modal/useModal'
import { useTablesStore } from '#stores/tablesStore'

import CreateTable from './create-table/CreateTable.vue'
import TableItem from './TableItem.vue'

import type { Table } from '#stores/tablesStore'

const { setComponent } = useModal()

const addTable = () => {
  setComponent(CreateTable)
}

const { getTables } = storeToRefs(useTablesStore())

const tables = computed(() =>
  getTables.value.map((table: Table) => ({
    id: table.id,
    name: table.name,
    moderatorCount: table.users.filter((user) => user.role === 'MODERATOR').length,
  })),
)
</script>

<style scoped>
.header {
  display: flex;
  gap: 20px;
  align-items: center;
  justify-content: space-between;
}

.list {
  display: flex;
  flex-flow: column;
  gap: 8px;
  height: 78%;
  padding: 0;
  overflow: auto;
  list-style: none;
}

.add-table {
  font-size: 14px;
  color: white !important;
  border-radius: 20px;
}
</style>
