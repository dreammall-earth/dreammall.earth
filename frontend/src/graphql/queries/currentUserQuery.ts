import { gql } from 'graphql-tag'

export const currentUserQuery = gql`
  query {
    currentUser {
      id
      name
      username
      room {
        id
        name
        public
        users {
          id
          role
          name
          username
        }
      }
    }
  }
`
