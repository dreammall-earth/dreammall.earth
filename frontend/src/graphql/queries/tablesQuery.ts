import { gql } from 'graphql-tag'

export const tablesQuery = gql(`{
  tables {
    id
    name
    public
    users {
      id
      name
      role
      username
    }
  }
}`)
