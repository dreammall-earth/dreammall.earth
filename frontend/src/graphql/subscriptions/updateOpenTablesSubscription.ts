import { gql } from 'graphql-tag'

export const updateOpenTablesSubscription = gql`
  subscription ($username: String!) {
    updateOpenTables(username: $username) {
      meetingID
      meetingName
      startTime
      participantCount
      attendees {
        fullName
      }
      joinLink
    }
  }
`
