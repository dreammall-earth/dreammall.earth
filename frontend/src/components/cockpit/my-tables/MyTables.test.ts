import { provideApolloClient } from '@vue/apollo-composable'
import { mount } from '@vue/test-utils'
import { createMockClient, createMockSubscription, IMockSubscription } from 'mock-apollo-client'
import { describe, expect, it, vi } from 'vitest'

import { currentUserQuery } from '#queries/currentUserQuery'
import { tablesQuery } from '#queries/tablesQuery'
import { updateOpenTablesSubscription } from '#subscriptions/updateOpenTablesSubscription'

import MyTables from './MyTables.vue'

const currentUserQueryMock = vi.fn()
const updateOpenTablesSubscriptionMock: IMockSubscription = createMockSubscription()
const tablesQueryMock = vi.fn()

const mockClient = createMockClient()

mockClient.setRequestHandler(
  currentUserQuery,
  currentUserQueryMock.mockResolvedValue({
    data: {
      currentUser: {
        id: 666,
        name: 'Current User',
        username: 'currentUser',
        table: null,
      },
    },
  }),
)

mockClient.setRequestHandler(updateOpenTablesSubscription, () => updateOpenTablesSubscriptionMock)

mockClient.setRequestHandler(
  tablesQuery,
  tablesQueryMock.mockResolvedValue({
    data: {
      tables: [
        {
          name: 'My Table',
          public: false,
          users: [
            {
              name: 'Current User',
              username: 'currentUser',
              role: 'MODERATOR',
            },
            {
              name: 'Bibi Bloxberg',
              username: 'bibi',
              role: 'VIEWER',
            },
            {
              name: 'Peter Lustig',
              username: 'peter',
              role: 'VIEWER',
            },
          ],
        },
      ],
    },
  }),
)

provideApolloClient(mockClient)

describe('MyTables', () => {
  const Wrapper = (props = {}) =>
    mount(MyTables, {
      props,
    })

  it('renders', () => {
    const wrapper = Wrapper()
    expect(wrapper.element).toMatchSnapshot()
  })
})
