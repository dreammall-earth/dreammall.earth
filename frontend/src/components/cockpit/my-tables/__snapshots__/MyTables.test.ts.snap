// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyTables > renders 1`] = `
<section
  class="cockpit-card"
  data-v-1451ba78=""
  data-v-c2e63817=""
>
  <div
    class="header"
    data-v-1451ba78=""
  >
    
    <div
      class="header"
      data-v-c2e63817=""
    >
      <h2
        data-v-c2e63817=""
      >
        $t('cockpit.myTables.header')
      </h2>
      <button
        class="add-table bg-primary px-3 py-1 d-flex align-center justify-center"
        data-v-c2e63817=""
      >
        <i
          aria-hidden="true"
          class="mdi mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default"
          data-v-c2e63817=""
        />
         $t('cockpit.myTables.addTable')
      </button>
    </div>
    
  </div>
  <hr
    class="hr"
    data-v-1451ba78=""
  />
  
  <div
    data-v-c2e63817=""
  >
    $t('cockpit.myTables.noTables')
  </div>
  
</section>
`;
